hydra-node/src/Hydra/Logging.hs

```diff
@@ -10,96 +10,78 @@ module Hydra.Logging (
   nullTracer,
   traceWith,
   traceInTVar,
-  LoggerName,
+  ToObject (..),
+  TracingVerbosity (..),
 
   -- * Using it
-  ToObject (..),
   Verbosity (..),
-  TracingVerbosity (..),
   withTracer,
   contramap,
   showLogsOnFailure,
 ) where
 
 import Hydra.Prelude
 
-import Cardano.BM.Backend.Switchboard (
-  Switchboard,
- )
-import Cardano.BM.Configuration.Static (
-  defaultConfigStdout,
- )
-import Cardano.BM.Data.LogItem (
-  LOContent (..),
-  LogObject (..),
-  LoggerName,
-  PrivacyAnnotation (..),
-  mkLOMeta,
- )
-import Cardano.BM.Data.Severity (
-  Severity (..),
- )
-import Cardano.BM.Data.Tracer (
-  ToObject (..),
-  TracingVerbosity (..),
- )
-import Cardano.BM.Setup (
-  setupTrace_,
-  shutdown,
- )
+import Cardano.BM.Tracing (ToObject (..), TracingVerbosity (..))
+import Control.Monad.Class.MonadAsync (async)
+import Control.Monad.Class.MonadFork (myThreadId)
+import Control.Monad.Class.MonadSTM (flushTBQueue, modifyTVar, newTBQueueIO, newTVarIO, readTVarIO, writeTBQueue)
+import Control.Monad.Class.MonadSay (MonadSay, say)
 import Control.Tracer (
   Tracer (..),
   natTracer,
   nullTracer,
   traceWith,
  )
+import Data.Aeson (encode, object, (.=))
+import qualified Data.ByteString.Lazy as LBS
+import qualified Data.Text as Text
+import qualified Data.Text.IO as TIO
 
-import qualified Cardano.BM.Configuration.Model as CM
-import qualified Cardano.BM.Data.BackendKind as CM
-import Control.Monad.Class.MonadSTM (modifyTVar, newTVarIO, readTVarIO)
-import Control.Monad.Class.MonadSay (MonadSay, say)
-
-data Verbosity = Quiet | Verbose LoggerName
+data Verbosity = Quiet | Verbose Text
   deriving (Eq, Show)
 
--- | Acquire a tracer that automatically shutdown once the action is done via
--- bracket-style allocation.
+defaultQueueSize :: Natural
+defaultQueueSize = 500
+
+-- | Start logging thread and acquire a 'Tracer'.
+-- This tracer will dump all messsages on @stdout@, one message per line,
+-- formatted as JSON.
 withTracer ::
   forall m msg a.
-  (MonadIO m, ToObject msg) =>
+  (MonadIO m, ToJSON msg) =>
   Verbosity ->
   (Tracer m msg -> IO a) ->
   IO a
-withTracer Quiet between = between nullTracer
-withTracer (Verbose name) between = do
-  bracket before after (between . natTracer liftIO . fst)
+withTracer Quiet action = action nullTracer
+withTracer (Verbose name) action = do
+  msgQueue <- newTBQueueIO defaultQueueSize
+  bracket_
+    (async $ logMessagesToStdout msgQueue)
```

Comment:
Slightly afraid about the wild async here (been bitten by that too often...). Since we're already in a bracket-style context within this `withTracer`, why not stick to `withAsync` ?

Reply:
I had a `withAsync` in a first iteration, but then I wanted to flush the log queue when an exception arises in the inner action hence the use of `bracket`. I will use the bracket inside the withAsync  which is safer.